datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java

```diff
@@ -1611,20 +1611,22 @@ void waitForStopToPropagateAndMarkDatastreamsStopped(List<DatastreamGroup> stopp
   private boolean markDatastreamsStopped(List<DatastreamGroup> stoppingDatastreamGroups, Set<String> failedStreams) {
     boolean success = true;
     boolean forceStop = _config.getForceStopStreamsOnFailure();
-    Set<String> stoppingStreams =
-        fetchDatastreamGroupsWithStatus(Collections.singletonList(DatastreamStatus.STOPPING)).
-            stream().flatMap(dg -> dg.getDatastreams().stream()).map(Datastream::getName).
-            collect(Collectors.toSet());
+
     for (DatastreamGroup datastreamGroup : stoppingDatastreamGroups) {
       for (Datastream datastream : datastreamGroup.getDatastreams()) {
         // Only streams that were confirmed to have stopped successfully will be transitioned to STOPPED state
-        if (stoppingStreams.contains(datastream.getName()) &&
-            (forceStop || !failedStreams.contains(datastream.getName()))) {
-          datastream.setStatus(DatastreamStatus.STOPPED);
-          _log.info("Transitioned datastream {} to STOPPED state", datastream.getName());
-          if (!_adapter.updateDatastream(datastream)) {
-            _log.error("Failed to update datastream: {} to stopped state", datastream.getName());
-            success = false;
+        if (forceStop || !failedStreams.contains(datastream.getName())) {
+          // fetch the latest from ZK, only STOPPING state datastreams will be marked as STOPPED
+          Datastream zkDatastream = _datastreamCache.getDatastream(datastream.getName(), true);
+          if (zkDatastream.getStatus() == DatastreamStatus.STOPPING) {
+            zkDatastream.setStatus(DatastreamStatus.STOPPED);
+            _log.info("Transitioning to STOPPED state dsName={}", zkDatastream.getName());
+            if (!_adapter.updateDatastream(zkDatastream)) {
+              _log.error("Failed to update to stopped state dsName={}", zkDatastream.getName());
+              success = false;
+            }
+          } else {
+            _log.info("Not in stopping state, dsName={} status={}", zkDatastream.getName(), zkDatastream.getStatus());
```

Comment:
nit: Make the logs consistent to have the state logged in upper case, "STOPPED" in lines 1625,1629.